File: meme_generator.go
// meme_generator.go
package main

import (
	"flag"
	"fmt"
	"image"
	"image/color"
	"image/draw"
	"image/jpeg"
	_ "image/png"
	"io/ioutil"
	"os"
	"path/filepath"
	"strconv"
	"strings"

	"github.com/fogleman/gg"
)

func parseHexColor(c string) color.RGBA {
	if len(c) == 7 && c[0] == '#' {
		r, _ := strconv.ParseInt(c[1:3], 16, 8)
		g, _ := strconv.ParseInt(c[3:5], 16, 8)
		b, _ := strconv.ParseInt(c[5:7], 16, 8)
		return color.RGBA{uint8(r), uint8(g), uint8(b), 255}
	}
	return color.RGBA{255, 255, 255, 255}
}

func memeGenerator(inputPath, outputPath, topText, bottomText string,
	fontSize float64, fontPath string, colorHex string,
	strokeColorHex string, strokeWidth float64, shadow bool, padding float64) error {
	img, err := gg.LoadImage(inputPath)
	if err != nil {
		return err
	}
	w, h := img.Bounds().Dx(), img.Bounds().Dy()
	dc := gg.NewContext(w, h)
	dc.DrawImage(img, 0, 0)

	// Шрифт
	if fontPath != "" {
		if err := dc.LoadFontFace(fontPath, fontSize); err != nil {
			// fallback
			dc.SetRGB(1, 1, 1)
		}
	} else {
		dc.SetRGB(1, 1, 1)
	}

	textColor := parseHexColor(colorHex)
	strokeColor := parseHexColor(strokeColorHex)

	// Верхний текст
	if topText != "" {
		dc.SetColor(textColor)
		dc.SetStrokeColor(strokeColor)
		dc.SetStrokeWidth(strokeWidth)
		dc.DrawStringAnchored(topText, float64(w)/2, padding, 0.5, 0)
	}

	// Нижний текст
	if bottomText != "" {
		dc.SetColor(textColor)
		dc.SetStrokeColor(strokeColor)
		dc.SetStrokeWidth(strokeWidth)
		dc.DrawStringAnchored(bottomText, float64(w)/2, float64(h)-padding, 0.5, 1)
	}

	// Сохранение
	return dc.SavePNG(outputPath)
}

func batchProcess(inputDir, outputDir, topText, bottomText string,
	fontSize float64, fontPath, colorHex, strokeColorHex string,
	strokeWidth float64, shadow bool, padding float64) error {
	os.MkdirAll(outputDir, 0755)
	exts := map[string]bool{".png": true, ".jpg": true, ".jpeg": true, ".bmp": true, ".tiff": true, ".webp": true}
	files, err := ioutil.ReadDir(inputDir)
	if err != nil {
		return err
	}
	var imageFiles []string
	for _, f := range files {
		if !f.IsDir() && exts[strings.ToLower(filepath.Ext(f.Name()))] {
			imageFiles = append(imageFiles, filepath.Join(inputDir, f.Name()))
		}
	}
	for i, f := range imageFiles {
		outPath := filepath.Join(outputDir, filepath.Base(f))
		fmt.Printf("[%d/%d] %s\n", i+1, len(imageFiles), filepath.Base(f))
		if err := memeGenerator(f, outPath, topText, bottomText, fontSize, fontPath,
			colorHex, strokeColorHex, strokeWidth, shadow, padding); err != nil {
			fmt.Printf("Error: %v\n", err)
		}
	}
	return nil
}

func main() {
	var (
		input      string
		output     string
		top        string
		bottom     string
		fontSize   float64
		fontPath   string
		color      string
		stroke     string
		strokeW    float64
		shadow     bool
		padding    float64
		batchDir   string
	)
	flag.StringVar(&input, "input", "", "Input image or dir (with --batch)")
	flag.StringVar(&output, "o", "meme.png", "Output file")
	flag.StringVar(&top, "t", "", "Top text")
	flag.StringVar(&bottom, "b", "", "Bottom text")
	flag.Float64Var(&fontSize, "s", 48, "Font size")
	flag.StringVar(&fontPath, "f", "", "Font path")
	flag.StringVar(&color, "c", "#FFFFFF", "Text color")
	flag.StringVar(&stroke, "stroke", "#000000", "Stroke color")
	flag.Float64Var(&strokeW, "stroke-width", 2, "Stroke width")
	flag.BoolVar(&shadow, "shadow", false, "Add shadow")
	flag.Float64Var(&padding, "padding", 20, "Padding")
	flag.StringVar(&batchDir, "batch", "", "Batch processing output dir")
	flag.Parse()

	if flag.NArg() > 0 {
		input = flag.Arg(0)
	}
	if input == "" {
		fmt.Println("Укажите входной файл или папку.")
		flag.Usage()
		os.Exit(1)
	}

	if batchDir != "" {
		if err := batchProcess(input, batchDir, top, bottom, fontSize, fontPath,
			color, stroke, strokeW, shadow, padding); err != nil {
			fmt.Println("Ошибка:", err)
			os.Exit(1)
		}
		fmt.Println("✅ Batch processing completed. Results in", batchDir)
	} else {
		if err := memeGenerator(input, output, top, bottom, fontSize, fontPath,
			color, stroke, strokeW, shadow, padding); err != nil {
			fmt.Println("Ошибка:", err)
			os.Exit(1)
		}
		fmt.Println("✅ Meme saved to", output)
	}
}

File: meme_generator.java
// meme_generator.java
import java.awt.*;
import java.awt.image.*;
import java.io.*;
import java.nio.file.*;
import java.util.*;
import javax.imageio.*;

public class meme_generator {
    public static void main(String[] args) throws Exception {
        String input = null, output = "meme.jpg";
        String topText = "", bottomText = "";
        int fontSize = 48;
        String fontPath = null, colorHex = "#FFFFFF", strokeHex = "#000000";
        float strokeWidth = 2;
        boolean shadow = false;
        int padding = 20;
        String batchDir = null;

        for (int i = 0; i < args.length; i++) {
            if (args[i].equals("-t") && i+1 < args.length) topText = args[++i];
            else if (args[i].equals("-b") && i+1 < args.length) bottomText = args[++i];
            else if (args[i].equals("-o") && i+1 < args.length) output = args[++i];
            else if (args[i].equals("-s") && i+1 < args.length) fontSize = Integer.parseInt(args[++i]);
            else if (args[i].equals("-c") && i+1 < args.length) colorHex = args[++i];
            else if (args[i].equals("-f") && i+1 < args.length) fontPath = args[++i];
            else if (args[i].equals("--stroke") && i+1 < args.length) strokeHex = args[++i];
            else if (args[i].equals("--stroke-width") && i+1 < args.length) strokeWidth = Float.parseFloat(args[++i]);
            else if (args[i].equals("--shadow")) shadow = true;
            else if (args[i].equals("--padding") && i+1 < args.length) padding = Integer.parseInt(args[++i]);
            else if (args[i].equals("--batch") && i+1 < args.length) batchDir = args[++i];
            else if (args[i].equals("-h") || args[i].equals("--help")) {
                System.out.println("Usage: meme_generator <input> [options]\n  -t <text>    Top text\n  -b <text>    Bottom text\n  -o <file>    Output\n  -s <N>       Font size\n  -c <hex>     Color\n  -f <file>    Font path\n  --stroke <hex>  Stroke color\n  --stroke-width <N>  Stroke width\n  --shadow     Add shadow\n  --padding <N>  Padding\n  --batch <dir>  Batch");
                return;
            } else if (input == null) input = args[i];
        }
        if (input == null) { System.err.println("Укажите входной файл или папку."); System.exit(1); }

        if (batchDir != null) {
            batchProcess(input, batchDir, topText, bottomText, fontSize, fontPath,
                         colorHex, strokeHex, strokeWidth, shadow, padding);
            System.out.println("✅ Batch processing completed. Results in " + batchDir);
        } else {
            memeGenerator(input, output, topText, bottomText, fontSize, fontPath,
                          colorHex, strokeHex, strokeWidth, shadow, padding);
            System.out.println("✅ Meme saved to " + output);
        }
    }

    static void memeGenerator(String inputPath, String outputPath, String topText, String bottomText,
                              int fontSize, String fontPath, String colorHex, String strokeHex,
                              float strokeWidth, boolean shadow, int padding) throws Exception {
        BufferedImage img = ImageIO.read(new File(inputPath));
        int w = img.getWidth(), h = img.getHeight();

        Graphics2D g = img.createGraphics();
        g.setRenderingHint(RenderingHints.KEY_TEXT_ANTIALIASING, RenderingHints.VALUE_TEXT_ANTIALIAS_ON);
        g.setRenderingHint(RenderingHints.KEY_RENDERING, RenderingHints.VALUE_RENDER_QUALITY);

        Color color = Color.decode(colorHex);
        Color stroke = Color.decode(strokeHex);

        Font font;
        if (fontPath != null && new File(fontPath).exists()) {
            font = Font.createFont(Font.TRUETYPE_FONT, new File(fontPath)).deriveFont(Font.BOLD, fontSize);
        } else {
            font = new Font("Impact", Font.BOLD, fontSize);
        }

        // Верхний текст
        if (topText != null && !topText.isEmpty()) {
            g.setFont(font);
            g.setColor(color);
            FontMetrics fm = g.getFontMetrics();
            int x = (w - fm.stringWidth(topText)) / 2;
            int y = padding + fm.getAscent();
            // Обводка
            g.setColor(stroke);
            g.setStroke(new BasicStroke(strokeWidth));
            for (int dx = -1; dx <= 1; dx++) {
                for (int dy = -1; dy <= 1; dy++) {
                    if (dx != 0 || dy != 0)
                        g.drawString(topText, x + dx, y + dy);
                }
            }
            g.setColor(color);
            g.drawString(topText, x, y);
        }

        // Нижний текст
        if (bottomText != null && !bottomText.isEmpty()) {
            g.setFont(font);
            g.setColor(color);
            FontMetrics fm = g.getFontMetrics();
            int x = (w - fm.stringWidth(bottomText)) / 2;
            int y = h - padding - fm.getDescent();
            g.setColor(stroke);
            g.setStroke(new BasicStroke(strokeWidth));
            for (int dx = -1; dx <= 1; dx++) {
                for (int dy = -1; dy <= 1; dy++) {
                    if (dx != 0 || dy != 0)
                        g.drawString(bottomText, x + dx, y + dy);
                }
            }
            g.setColor(color);
            g.drawString(bottomText, x, y);
        }

        g.dispose();
        ImageIO.write(img, "jpg", new File(outputPath));
    }

    static void batchProcess(String inputDir, String outputDir, String topText, String bottomText,
                             int fontSize, String fontPath, String colorHex, String strokeHex,
                             float strokeWidth, boolean shadow, int padding) throws Exception {
        Files.createDirectories(Paths.get(outputDir));
        String[] exts = {".png", ".jpg", ".jpeg", ".bmp", ".tiff", ".webp"};
        Set<String> extSet = new HashSet<>(Arrays.asList(exts));
        File[] files = new File(inputDir).listFiles();
        if (files == null) return;
        int total = 0;
        java.util.List<File> imageFiles = new ArrayList<>();
        for (File f : files) {
            if (f.isFile() && extSet.contains(f.getName().substring(f.getName().lastIndexOf('.')).toLowerCase())) {
                imageFiles.add(f);
            }
        }
        total = imageFiles.size();
        for (int i = 0; i < total; i++) {
            File f = imageFiles.get(i);
            String outPath = outputDir + File.separator + f.getName();
            System.out.println("[" + (i+1) + "/" + total + "] " + f.getName());
            memeGenerator(f.getPath(), outPath, topText, bottomText, fontSize, fontPath,
                          colorHex, strokeHex, strokeWidth, shadow, padding);
        }
    }
}
